### PR TITLE
Add Tracing context in Key Vault keys package

### PR DESCRIPTION
**Pull request by samvaity**
- Added tracing context to Keyvault keys package
- Added `T` with `Response<T>` "overload in sync and async Key vault keys client.
- Updated tests if necessary to work with the new return types.

Closes: #4475

### CODE REVIEW

File: eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml [hunk 1]
```diff
@@ -443,6 +443,15 @@
     <Bug pattern="NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE"/>
   </Match>
 
+  <!-- Incorrect flagging, if the response is null a StorageException would have been thrown -->
+  <Match>
+    <Or>
+      <Class name="com.azure.security.keyvault.keys.KeyClient"/>
+    </Or>
+    <Bug pattern="NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE"/>
```
Comment (samvaity) [inline eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml]: @JonathanGiles 
Was getting this on all the T return type methods.
Checked with @g2vinay also on this, he said we ideally shouldn't be getting anything like this.
Added this as a temporary fix, would you know more about it.

File: sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/KeyClientJavaDocCodeSnippets.java [hunk 2]
```diff
@@ -89,9 +100,337 @@ public KeyAsyncClient createAsyncClientWithPipeline() {
...
+        EcKeyCreateOptions ecKeyCreateOptions = new EcKeyCreateOptions("keyName")
+            .curve(KeyCurveName.P_384)
+            .notBefore(OffsetDateTime.now().plusDays(1))
+            .expires(OffsetDateTime.now().plusYears(1));
+        Key ecKey = keyClient.createEcKey(ecKeyCreateOptions);
+        System.out.printf("Key is created with name %s and id %s \n", ecKey.name(), ecKey.id());
+        // END: com.azure.keyvault.keys.keyclient.createEcKey#keyOptions
+
+        // BEGIN: com.azure.keyvault.keys.keyclient.importKey#keyImportOptions
+        // JsonWebKey jsonWebKeyToImport;
```
Comment (samvaity) [inline sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/KeyClientJavaDocCodeSnippets.java]: These code snippets would need a valid `jsonWebKeyToImport`  to work.
@g2vinay do we have a good working example for this or should we remove this for now?

File: eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml (hunk 1 above)
Comment (JonathanGiles) [reply eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml]: Let's sync up sometime to discuss exactly where this issue is arising and how we should resolve.

File: sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java [hunk 3]
```diff
@@ -94,11 +94,36 @@
      * @param keyType The type of key to create. For valid values, see {@link KeyType KeyType}.
      * @throws ResourceModifiedException if {@code name} or {@code keyType} is null.
      * @throws HttpRequestException if {@code name} is empty string.
+     * @return A {@link Mono} containing the {@link Key created key}.
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java]: I see an old code sample above. I assume you'll update these so I won't point them out any longer in this PR review.

File: sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java [hunk 4]
```diff
@@ -94,11 +94,36 @@
...
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Key> createKey(String name, KeyType keyType) {
+        return withContext(context -> createKey(name, keyType, Context.NONE))
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java]:
```suggestion
        return withContext(context -> createKey(name, keyType, context))
```

File: sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java [hunk 5]
```diff
@@ -135,11 +160,12 @@
...
      * @throws HttpRequestException if {@code name} is empty string.
-     * @return A {@link Mono} containing a {@link Response} whose {@link Response#value() value} contains the {@link Key created key}.
+     * @return A {@link Mono} containing the {@link Key created key}.
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
-    public Mono<Response<Key>> createKey(KeyCreateOptions keyCreateOptions) {
-        return withContext(context -> createKey(keyCreateOptions, context));
+    public Mono<Key> createKey(KeyCreateOptions keyCreateOptions) {
+        return withContext(context -> createKey(keyCreateOptions, context))
+            .flatMap(keyResponse -> Mono.justOrEmpty(keyResponse.value()));
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java]: Not a fan of all the same lines of code for flatMap. I would suggest moving this out to a utility method in azure-core, and then doing something like `.flatMap(FluxUtils::flattenResponse)`, where flattenResponse would be the code you have above.

File: sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyClientBuilder.java [hunk 6]
```diff
@@ -124,6 +125,7 @@ public KeyAsyncClient buildAsyncClient() {
         policies.add(retryPolicy);
         policies.add(new BearerTokenAuthenticationPolicy(credential, KeyAsyncClient.KEY_VAULT_SCOPE));
         policies.addAll(this.policies);
+        HttpPolicyProviders.addAfterRetryPolicies(policies);
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyClientBuilder.java]: There should also be a 'beforeRetryPolicies' line that you add.

File: sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/KeyClientJavaDocCodeSnippets.java (hunk 2 above)
Comment (g2vinay) [reply sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/KeyClientJavaDocCodeSnippets.java]: We will have a good working example with Crypto API implementation.
Where we can import local crypto keys into the service. I'll let you know, once I add that.
For now, let's keep it in the code files. Add a TODO there.

**Comment by danieljurek** [comment]
[Following up on CI build that was improperly triggered]

Please ignore the `java - cosmos - ci` build failure. This was triggered by a bug. The bug has been fixed (https://github.com/Azure/azure-sdk-for-java/pull/4520) so we shouldn't see this problem going forward.

**Comment by JonathanGiles** [comment]
The one checkstyle error is the following: 

`[ERROR] /azure-sdk-for-java/sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/KeyClientJavaDocCodeSnippets.java:16:8: Unused import - com.azure.security.keyvault.keys.models.KeyImportOptions. [UnusedImports]`

File: eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml (hunk 1 above)
Comment (samvaity) [reply eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml]: Keeping this fix, for now.
 Opened an [issue](https://github.com/azure/azure-sdk-for-java/issues/4556) for future investigation.

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/FluxUtil.java [hunk 7]
```diff
@@ -199,6 +200,21 @@ public static boolean isFluxByteBuf(Type entityType) {
...
+     *<p>
+     * If the response {@link Response#value() value} is empty, an empty Mono is returned.
+     *</p>
+     * <p><strong>Code samples</strong></p>
+     * {@codesnippet com.azure.core.implementation.util.fluxutil.toMono}
+     *
+     * @param response whose {@link Response#value() value} is to be converted
+     * @return The converted {@link Mono}
+     */
+    public static <T> Mono<T> toMono(Response<T> response) {
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/FluxUtil.java]: Somewhat redundant JavaDoc (and especially the code snippet) as this is an implementation class and therefore will never be generated.

File: sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java [hunk 8]
```diff
@@ -122,24 +144,19 @@
...
      * @throws ResourceModifiedException if {@code keyCreateOptions} is malformed.
      * @throws HttpRequestException if {@code name} is empty string.
-     * @return A {@link Mono} containing a {@link Response} whose {@link Response#value() value} contains the {@link Key created key}.
+     * @return A {@link Mono} containing the {@link Key created key}.
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
-    public Mono<Response<Key>> createKey(KeyCreateOptions keyCreateOptions) {
-        return withContext(context -> createKey(keyCreateOptions, context));
+    public Mono<Key> createKey(KeyCreateOptions keyCreateOptions) {
+        return withContext(context -> createKey(keyCreateOptions, context))
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java]: Can't you call createKeyWithResponse here?

File: sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java [hunk 9]
```diff
@@ -169,15 +185,33 @@
...
+     *
+     * @param rsaKeyCreateOptions The key configuration object containing information about the rsa key being created.
+     * @throws NullPointerException if {@code rsaKeyCreateOptions} is {@code null}.
+     * @throws ResourceModifiedException if {@code rsaKeyCreateOptions} is malformed.
+     * @throws HttpRequestException if {@code name} is empty string.
+     * @return A {@link Mono} containing the {@link Key created key}.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Key> createRsaKey(RsaKeyCreateOptions rsaKeyCreateOptions) {
+        return withContext(context -> createRsaKey(rsaKeyCreateOptions, context))
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyAsyncClient.java]: Same here

File: sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/ListOperationsAsync.java [hunk 10]
```diff
@@ -35,11 +35,11 @@ public static void main(String[] args) throws InterruptedException {
         keyAsyncClient.createEcKey(new EcKeyCreateOptions("CloudEcKey")
                 .expires(OffsetDateTime.now().plusYears(1)))
                 .subscribe(keyResponse ->
-                        System.out.printf("Key is created with name %s and type %s \n", keyResponse.value().name(), keyResponse.value().keyMaterial().kty()));
+                        System.out.printf("Key is created with name %s and type %s \n", keyResponse.name(), keyResponse.keyMaterial().kty()));
 
         Thread.sleep(2000);
 
-        keyAsyncClient.createRsaKey(new RsaKeyCreateOptions("CloudRsaKey")
+        keyAsyncClient.createRsaKeyWithResponse(new RsaKeyCreateOptions("CloudRsaKey")
```
Comment (g2vinay) [inline sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/ListOperationsAsync.java]: We should probably show a detailed sample for withResponse, processing/printing the headers info.

**Comment by g2vinay** [comment]
We should add one more test case for withResponse, parsing response headers and validating their values. But don't worry about it in this PR. as it will require recording it too. 
I'll add that in my Keys crypyo PR.

File: sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/ListOperationsAsync.java (hunk 10 above)
Comment (samvaity) [reply sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/ListOperationsAsync.java]: But that might lead to a lot of similar-looking samples 🤷‍♀

Comment (g2vinay) [reply sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/ListOperationsAsync.java]: We can just update the existing sample, and make it print the response headers info along with the returned key details.

Comment (samvaity) [reply sdk/keyvault/azure-keyvault-keys/src/samples/java/com/azure/security/keyvault/keys/ListOperationsAsync.java]: I think we should have samples which show the T response type methods as that is the more used one. I will update these examples in #4549 to use T response methods.

**Pull request merged**